[2주차 과제] Spring Web 으로 ToDo REST API 만들기

jdalma opened the pull request:
### REST API

- ToDo 목록 얻기 - `GET /tasks`
- ToDo 상세 조회하기 - `GET /tasks/{id}`
- ToDo 생성하기 - `POST /tasks`
- ToDo 제목 수정하기 - `PUT/PATCH /tasks/{id}`
- ToDo 삭제하기 - `DELETE /tasks/{id}`

한 주간 잘 부탁드립니다 ㅎㅎ

jdalma committed:
feat: TaskController 추가

jdalma committed:
chore: Spring Dev Tools 추가

jdalma committed:
feat: 화면 추가

jdalma committed:
feat: CRUD 추가

jdalma committed:
fix: 네트워크 메소드 수정

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+@Service
+public class TaskService {
```
TaskService에서 제공해야하는 기능들을 인터페이스로 만들어서 구현과 명세를 분리해보세요!

jdalma replied:
일단 `TaskService`만을 위한 `TaskServiceInterface`를 추가해 보았습니다 [7d74139](https://github.com/CodeSoom/spring-week2-assignment-1/pull/94/commits/7d74139ebbefc9e041cd41672c1700c13171f679)

<br>

현재 필요한 인터페이스는 CRUD의 명세만 가지고 있으면 되기 때문에 `CRUDInterface`를 만들면 어떨까 고민해 보았습니다
하지만 `service`가 관심가지고 있는 `model`을 어떻게 알려줘야할지 몰라서 위와 같이 작성하였습니다

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,43 @@
+@Service
+public class TaskService {
+    private Map<Long , Task> tasks = new ConcurrentHashMap<>();
+    private AtomicLong taskId = new AtomicLong(1L);
```
Id를 생성하는 기능을 분리해보세요!

jdalma replied:
`TaskIdGenerator`라는 클래스를 Bean으로 등록하여 `service`에서 주입받아 보았습니다

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,53 @@
+
+    @RequestMapping(path = "/{taskId}" , method = {RequestMethod.PUT , RequestMethod.PATCH})
+    public void updateTask(@PathVariable long taskId ,
+                           @RequestBody Task newTask){
```
`newTask` 는 항상 id가 없습니다. `Task` 를 사용하면 id가 항상 null 이므로 쓸모가 없는 필드가 있는 느낌이에요!

`updateTask` 만을 위한 Dto 객체를 만들어보면 어떨까요?!

jdalma replied:
DTO에 대해 듣기만해서 [정리](https://jdalma.github.io/docs/algorithmTheory/dto/) 해보는 시간을 가져봤습니다 
`TaskMapper`를 추가하여 적용해보았습니다 [89790b6](https://github.com/CodeSoom/spring-week2-assignment-1/pull/94/commits/89790b64cc4d106aaa7ba69a88b738ca3d35b73b)

<br>

> 도메인 필드가 더 추가되고 화면에서 담겨 올 정보들이 변경된다면 **변경사항에 맞는 DTO**를 추가한다
> 해당 DTO를 도메인으로 변경해주는 메서드가 추가로 필요하다면 `mapper`에 메서드를 추가한다  

라고 이해하였는데 맞을까요??

jdalma committed:
refactor: Setter제거 , 필드 private로 변경

jdalma committed:
TaskDTO , TaskMapper 추가 및 적용

jdalma committed:
TaskService 구현과 명세를 분리 , TaskId 관리를 분리

jdalma committed:
toString 추가

jdalma committed:
Optional 반환 , 404 응답 코드 추가

jdalma committed:
Optional 제거 , null 비교로 수정함

[단지 값을 얻을 목적이라면 Optional 대신 null 비교](http://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/)를 보고 수정하였습니다

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/mappers/TaskMapper.java:
```diff
@@ -0,0 +1,13 @@
+import com.codesoom.assignment.models.TaskDTO;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TaskMapper {
+
+    public Task toNewTask(Long taskId , TaskDTO taskDTO){
+        return new Task(taskId , taskDTO.getTitle());
+    }
+}
```
좋습니다! `TaskMapper` 를 컴포넌트화했다면 TaskMapper는 어느 모듈에 있어야 할까요?
서비스일까요? 컴트롤러일까요?
한번 고민해보아요!

jdalma replied:
**컨트롤러**는 웹 관련(요청과 응답)에 관한 책임만 있다고 생각하여 `TaskDTO`의 객체를 받아냈다면 **컨트롤러**의 역할은 다 해냈다고 생각하였습니다
그래서 **서비스**에서 주입받아 사용하였습니다

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/models/TaskDTO.java:
```diff
@@ -0,0 +1,17 @@
+public class TaskDTO {
+    private String title;
+
+    public TaskDTO() {
+    }
```
기본생성자를 없애보는 시간을 가져요! 그렇게 된다면 `title`에 `final` 을 붙일 수 있게 되요!
`final` 을 붙이게 된다는 것은 `taskDTO` 가 한번 생성되면 그 값은 변할리 없다는 것을 보증하죠!

jdalma replied:
Jackson의 ObjectMapper가 기본 생성자를 사용하는 줄 알고있어 불가능할 줄 알았는데 , 이번에 **JSON Annotation**에 대해 알게 되었습니다 ㅎㅎ

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskIdGenerator.java:
```diff
@@ -0,0 +1,15 @@
+    public Long getNextId(){
+        return taskId.getAndIncrement();
+    }
+}
```
ID를 생성하는 것은 꼭 `Task` 만을 위한 것이 아닐 것입니다!

유저를 만들 때도 ID 생성기가 필요할 수도 있겠죠! 

그렇다면 `ID를 생성하는 역할` 을 가진 인터페이스를 만들어 볼 수 있을거에요!

jdalma replied:
`IdGenerator` 인터페이스를 추가해 봤습니다 [e14986e](https://github.com/CodeSoom/spring-week2-assignment-1/pull/94/commits/e14986e1379e378528646b1ab7d2482296cb82d6)

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskServiceInterface.java:
```diff
@@ -0,0 +1,16 @@
+    List<Task> getAllTask();
+    Task getTaskById(Long id);
+    void createTask(TaskDTO taskDTO);
+    Task updateTask(Long id , TaskDTO taskDTO);
```
컨트롤러는 보니 리턴값이 null 인 경우가 있군요. 이 경우에 대한 주석이 있으면 좋겠어요!

jdalma committed:
IdGenerator 인터페이스 추가

jdalma committed:
TaskDTO 필드 private로 수정 , 기본 생성자 제거 , JSON 어노테이션 사용

jdalma committed:
IdGenerator 인터페이스의 메서드 longTypeGenerate를 generate로 변경

반환값을 보면 타입을 알 수 있으므로 메서드 명에 타입을 제거

jdalma committed:
TaskServiceInterface void메서드 주석 추가

jdalma committed:
다형성

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/IdGenerator.java:
```diff
@@ -0,0 +1,6 @@
+package com.codesoom.assignment.services;
+
+public interface IdGenerator {
+
+    Long generate();
+}
```
너무 좋아요! 하지만 아주 필수적인 코드가 빠졌네요! 주석이요! 문서를 잘 적어주세요! 

또한 인터페이스는 이름과는 다르게 Long 만을 생성하는 인터페이스네요! 

이것을 좀 더 범용으로 바꿔볼 수는 없을까요?

jdalma replied:
`service`가 `IdGenerator`를 구현할 때 `id`타입을 지정할 수 있게 해줬습니다

[003956a](https://github.com/CodeSoom/spring-week2-assignment-1/pull/94/commits/003956ab8f7cbf5bd0a4d8db0c1fa243c4779fca)

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskServiceInterface.java:
```diff
@@ -0,0 +1,23 @@
+     * @param taskDTO
+     * @return null
+     */
+    void createTask(TaskDTO taskDTO);
```
아무것도 리턴하지 않는 것과 `null` 을 리턴하는 것은 차이가 있죠!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskServiceInterface.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.List;
+import java.util.Optional;
+
+public interface TaskServiceInterface {
```
인터페이스의 메소드에는 모두 주석이 있어야합니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Override
+    public List<Task> getAllTask(){
+        return new ArrayList<>(tasks.values());
```
task.values 를 그대로 리턴하는 것과 어떻게 다른 걸까요?!

jdalma replied:
`values()` 메서드는 `Collection<V>`타입으로 반환 

```java
public Collection<V> values() {
    ValuesView<K,V> vs;
    if ((vs = values) != null) return vs;
    return values = new ValuesView<K,V>(this);
}
```

List Interface가 Collection을 확장하고 있어서 `new ArrayList`로 공간을 새로 만들지 않고
반환 타입을 `Collection<V>`로 수정하였습니다

```java
public interface List<E> extends Collection<E>
```

ssisksl77 replied:
좋습니다. `List` 는 `Collection` 과 어떻게 다른지 공유해보면 좋은 경험일 것 같아요!

jdalma committed:
IdGenerator 인터페이스 제네릭 추가 generate 메소드 반환 타입 수정

jdalma committed:
TaskServiceInterface를 CRUDInterface로 변경

jdalma committed:
TaskServiceInterface를 CRUDInterface로 변경

jdalma commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.concurrent.ConcurrentHashMap;
+
+@Service
+public class TaskService implements CRUDInterface<TaskDTO , Task>{
```
`DTO`나 `model`이 더 늘어나게 되면 `DTO`와 `model`을 감쌀 수 있는 방법이 추가로 필요할 것 같지만...
한번 수정해 보았습니다

jdalma committed:
IdGenerator 주석 추가

jdalma committed:
IdGenerator 메서드 주석 추가

jdalma committed:
반환 Collection<Task>로 수정

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/CRUDInterface.java:
```diff
@@ -0,0 +1,40 @@
+     * @param id Collection에서 가져올 id 입니다
+     * @return id에 해당하는 Out타입의 객체를 반환합니다
+     */
+    Out selectById(Long id);
```
멋진 시도를 하고 계세요!

범용 인터페이스를 만드는 것은 항상 어렵습니다. 

만약 이것이 Task에 관련없는 범용 인터페이스라면 일단은 `services` 패키지에 없어도 되지 않나? 
라는 생각이 들기도 하네요! `IdGenerator` 도 마찬가지고요! (이것이 정답은 아니에요!)

추가로 만약 이것이 Task와 관련없는 범용이라면, `id` 가 `Long` 이라는 가정은 위험할 것 같아요!

jdalma replied:
`public interface CRUDInterface<In , Out , K>` id의 타입도 지정할 수 있게 변경해 보았습니다

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/CRUDInterface.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.Collection;
+
+/**
+ * {@link java.util.Collection}을 조작하는 클래스에 의해 구현되는 인터페이스 입니다
```
이것이 무엇을 뜻하는 것인지 설명해주세요!

`Collection` 은 현재 `selectAll` 에서만 사용하는 것 같아서 헷갈립니다!

jdalma replied:
"`Collection`을 구현하는 **자료 구조**들을 조작하는 클래스" 라고 전달하고 싶었습니다

"Domain을 조작하는 클래스"로 수정해 보았습니다

jdalma committed:
역할에 맞게 패키지 분리

jdalma committed:
CRUDInterface id 타입 지정

jdalma committed:
주석 내용 수정

jdalma committed:
@ExceptionHandler , @RestControllerAdvice 적용

DataNotFoundException 커스텀 예외 추가

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.concurrent.ConcurrentHashMap;
+
+@Service
+public class TaskService implements CRUDInterface<TaskDTO , Task , Long> {
```
TaskService에 `TaskDTO` 가 있는 것은 좋은 방법이 아닐 수 있습니다!

어떤 문제가 야기될 지 한번 생각해보는 시간을 가져봅시다!

jdalma replied:
지금 상황으로는 `CRUDInterface`때문에 `DTO`클래스가 더 늘어난다면 타입을 더 추가해 줄 방법을 찾아야 할 것 같습니다
떠오르는 해결방법은 부모 클래스나 추가적인 인터페이스를 사용해서 감싸줘야할 것 같습니다.. 하지만 굉장히 비효율적일 것 같네요 ㅎㅎ

ssisksl77 replied:
TaskInterface는 MVC 패턴의 관점에서보면 서비스 레이어에 존재해야 하는 인터페이스이므로, 컨트롤러 레이어에 있는 객체를 알고 있는 것은 좋지 않습니다!
그러므로 `DTO` 와 독립적인 객체가 있어야 할 것입니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,54 @@
+    @Override
+    public Task insert(TaskDTO taskDTO){
+        Long nextId = gen.generate();
+        Task newTask = mapper.toNewTask(nextId , taskDTO);
```
좋습니다!

```java
new Task(taskId , taskDTO.getTitle())
```

와 비교해서 어떤 장점이 있는 것일까요? 
제가 만약 여기서 "`new Task` 를 직접 쓰는 것이 직관적이지 않나요?" 라고 코드리뷰 중에 질문을 남겼다고 가정해봅시다!

jdalma replied:
`new Task`를 쓰는 것이 직관적인 것 같습니다
하지만 지금은 `TaskDTO` 한 개를 사용하지만 `Task`가 더 복잡해지거나 `DTO`클래스가 늘어난다면 한 눈에 보기가 어려울 것 같은 생각이들어 `DTO`와 `Domain` 사이를 연결해주는 `mapper`를 두는것이 좋다고 판단했습니다

ssisksl77 replied:
좋습니다! 

한가지 아쉬운 점은 mapper가 모든 Dto를 알아야 한다는 점입니다. 각 Dto에 Task를 만드는 메소드가 있으면 될 것 같아요!

jdalma replied:
아 그런 방법이 있네요 ㅎㅎㅎ 감사합니다 !!!

jdalma committed:
Update시 해당 id의 정보가 없다면 early return

jdalma committed:
ErrorResult클래스 추가

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -42,6 +42,9 @@ public Task insert(TaskDTO taskDTO){
 
     @Override
     public Task update(Long id, TaskDTO taskDTO) {
+        if(tasks.get(id) == null){
+            return null;
+        }
```
`CRUDInterface` 의 모든 `update` 는 `id` 를 못찾았을 때, null을 리턴해야하는 것이라면 `CRUDInterface` 문서에 관련 설명을 추가하는 것이 좋습니다.

jdalma replied:
아 넵 ㅎㅎ
이럴 때 `Optional`을 사용하는건 어떨까요??

ssisksl77 replied:
Optional을 리턴하는 것도 나쁘지 않습니다. 하지만 그럼에도 문제가 완벽하게 해결되지는 않습니다. 

`optional` 값이 비어있을 때 `update` 를 실패한 것인지 `update` 할 Task를 못 찾았다는 것인지는 명확하게 문서로 남겨놓는 것이 좋습니다.
